src/vs/editor/common/modes.ts

```diff
@@ -392,7 +392,7 @@ export interface ReferenceProvider {
 	/**
 	 * Provide a set of project-wide references for the given position and document.
 	 */
-	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, token: CancellationToken): Location[] | Thenable<Location[]>;
+	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, progress: (value: Location) => void, token: CancellationToken): Location[] | Thenable<Location[]>;
```

Comment:
My first solution used `PPromise`, but one piece of feedback from the vscode team was to avoid that:

> Wrt the API design the Thenable isn't what I prefer. The Thenable is the common denominator between various promise implementations. I know that there are some implementations, like WinJS.Promise, that have a progress callback but in the end it's non standard and we don't want to push that. Instead, something like we have explored for #18066 should be used. There is some proposed API that defines Progress<T>

Specific answers to your questions that I will add to the documentation as well

1. Progress has type Location (not Location[]) so it is only one result. One reason I chose this signature was so it wouldn't be possible to confuse the progress callback as returning "all results up until this point". That said, if a language server may return multiple results at once, it might be nice for this progress to support sending more than one result at a time (to minimize UI refresh). In either case, I think the intent would be for progress to be incremental, and I will add doc to that effect.
2. No, will document.
3. Progress callbacks after the promise has resolved will be ignored. Will document.